lib/experimental/class-wp-theme-json-resolver-gutenberg.php

```diff
@@ -98,4 +98,47 @@ public static function get_theme_data( $deprecated = array(), $settings = array(
 		return $with_theme_supports;
 	}
 
+	public static function get_block_data() {
+		$registry = WP_Block_Type_Registry::get_instance();
+		$blocks   = $registry->get_all_registered();
+		$config   = array( 'version' => 1 );
+		foreach ( $blocks as $block_name => $block_type ) {
+			if ( isset( $block_type->supports['blockStyles'] ) ) {
+				$config['styles']['blocks'][ $block_name ] = static::removeJsonComments( $block_type->supports['blockStyles'] );
+			}
+		}
+
+		// Core here means it's the lower level part of the styles chain.
+		// It can be a core or a third-party block.
+		return new WP_Theme_JSON( $config, 'core' );
+	}
+
+	private static function removeJsonComments( $array ) {
+		unset( $array['//'] );
+		foreach ( $array as $k => $v ) {
+			if ( is_array( $v ) ) {
+				$array[ $k ] = static::removeJsonComments( $v );
+			}
+		}
+
+		return $array;
+	}
+
+	public static function get_merged_data( $origin = 'custom' ) {
+		if ( is_array( $origin ) ) {
+			_deprecated_argument( __FUNCTION__, '5.9' );
+		}
+
+		$result = new WP_Theme_JSON_Gutenberg();
+		$result->merge( static::get_core_data() );
+		$result->merge( static::get_block_data() );
+		$result->merge( static::get_theme_data() );
```

Comment:
This is looking better. Core is the lowest level, then what the blocks add and finally what the Theme defines.

As long as, for blocks opting into this functionality, the ordering of the output CSS is _the same_ as it was prior to this PR then we should be good.

Reply:
This reminds me of my own PR [here](https://github.com/WordPress/gutenberg/pull/40656#discussion_r861631990) where the order of merging settings seems to be weird.

Reply:
> ...seems to be weird.

Can you clarify whether you are raising a concern here that should be addressed? 🙇

Reply:
I don't think I had a concern ..